source/kernel/src/services/sdmmc.rs

```diff
@@ -0,0 +1,467 @@
+                rsp_crc: true,
+                buffer: None,
+            })
+            .await?
+        {
+            Response::Short { value, .. } => {
+                tracing::trace!("CMD8 response: {value:#x}");
+                if value == 0x1AA {
+                    card_type = CardType::SD2;
+                }
+            }
+            Response::Long(_) => return Err(Error::Response),
+        }
+
+        // TODO: limit the number of attempts
+        let ocr = loop {
+            // Go to *APP* mode before sending application command
+            self.cmd(Command {
+                index: 55,
+                argument: 0,
+                options: HardwareOptions::None,
+                kind: CommandKind::Control,
+                rsp_type: ResponseType::Short,
+                rsp_crc: true,
+                buffer: None,
+            })
+            .await?;
+
+            let mut op_cond_arg = OCR_VOLTAGE_MASK & 0x00ff8000;
+            if card_type != CardType::SD1 {
+                op_cond_arg = OCR_HCS | op_cond_arg;
+            }
+            match self
+                .cmd(Command {
+                    index: 41,
+                    argument: op_cond_arg,
+                    options: HardwareOptions::None,
+                    kind: CommandKind::Control,
+                    rsp_type: ResponseType::Short,
+                    rsp_crc: false,
+                    buffer: None,
+                })
+                .await?
+            {
+                Response::Short { value, .. } => {
+                    tracing::trace!("ACMD41 response: {value:#x}");
+                    if value & OCR_NBUSY == OCR_NBUSY {
+                        // Card has finished power up, data is valid
+                        break value;
+                    }
+                }
+                Response::Long(_) => return Err(Error::Response),
+            }
+
+            // TODO: wait some time (e.g., 1ms?)
+        };
+
+        if (ocr & OCR_HCS) == OCR_HCS {
+            card_type = CardType::SDHC;
+        }
+
+        Ok(card_type)
+    }
+
+    /// Get the card identification register
+    pub async fn get_cid(&mut self) -> Result<CardIdentification, Error> {
+        match self
+            .cmd(Command {
+                index: 2,
+                argument: 0,
+                options: HardwareOptions::None,
+                kind: CommandKind::Control,
+                rsp_type: ResponseType::Long,
+                rsp_crc: true,
+                buffer: None,
+            })
+            .await?
+        {
+            Response::Short { .. } => return Err(Error::Response),
+            Response::Long(value) => {
+                tracing::trace!("CMD2 response: {value:?}");
+                // TODO: map [u32; 4] value to u128
+                Ok(CardIdentification(0))
+            }
+        }
+    }
+
+    /// Get the relative card address
+    pub async fn get_rca(&mut self) -> Result<u32, Error> {
```

Comment:
i wonder if we should consider adding a `Rca(u32)` or `RelativeCardAddress(u32)` newtype, so that the user can only call methods like `select` and, `set_wide_bus`, `read`, etc, with real RCAs they obtained from this method, rather than random `u32`s that might not index a real card?

Reply:
Good idea!